src/trajectory.jl

```diff
@@ -227,22 +227,52 @@ samplecand(rng, τ::StaticTrajectory{EndPointTS}, h, z) = step(τ.integrator, h,
 
 ### Multinomial sampling from trajecory
 
-function randcat(rng::AbstractRNG, xs, p)
-    u = rand(rng)
+function randcat(rng::AbstractRNG, p_unorm::AbstractVector{T}) where {T}
+    p = p_unorm ./ sum(p_unorm)
+    u = rand(rng, T)
     cp = zero(eltype(p))
     i = 0
     while cp < u
-        cp += p[i +=1]
+        cp += p[i+=1]
     end
-    return xs[max(i, 1)]
+    return max(i, 1)
+end
+
+function randcat(rng, P_unorm::AbstractMatrix{T}) where {T}
+    P = P_unorm ./ sum(P_unorm; dims=2)
+    u = rand(rng, T, size(P, 1))
+    C = cumsum(P; dims=2)
+    is = convert.(Int, vec(sum(C .< u; dims=2)))
+    return max.(is, 1)
+end
+
+randcat(rng::AbstractRNG, xs, p::AbstractVector) = xs[randcat(rng, p)]
```

Comment:
Maybe change `xs` into `zs` for consistency? Also adding type annotation makes it more readable here.
```suggestion
randcat(rng::AbstractRNG, zs::Vector{PhasePoint}, p::AbstractVector) = xs[randcat(rng, p)]
```